### ISSUE TITLE
[Git-bash (cygwin)] Progress output is duplicated in new lines

### ISSUE THREAD

**Issue by zlorf** [rclone/rclone#4415]
#### What is the problem you are having with rclone?

The progress is printed in new lines, instead of overwriting the previous lines. See screenshot:

![rclone-progress-error](https://user-images.githubusercontent.com/1769916/86849173-9469e880-c0af-11ea-8570-3b2bc9d50f1b.png)

Here it's not so bad, but imagine when uploading several files, it quickly clobbers all the terminal scroll history.

#### What is your rclone version (output from `rclone version`)

rclone v1.52.2
- os/arch: windows/amd64
- go version: go1.14.4

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

Windows 10, 64 bit
With git-bash (cygwin / mintty).
The problem DOES NOT reproduce with git-cmd.

####  Which cloud storage system are you using? (eg Google Drive)

B2

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

rclone copy

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

Not relevant

**Comment by ncw** [comment]
Does it work if you make the terminal window a bit wider?

Do you have an environmental variable called `TERM` set? If so can you try unsetting it or setting it if you don't?

**Comment by zlorf** [comment]
Changing terminal width does not solve the problem.
TERM is set to `xterm`

**Comment by ncw** [comment]
If you try rclone ncdu does that work?

Assuming that rclone is outputting the vt100 command codes then this might be a bug in git bash.

Rclone uses only a very small number of VT control codes.

**Comment by zlorf** [comment]
There is no output at all from `ncdu` command. It works properly in git-cmd.
In this case, it seems like a problem in git-bash after all.

**Comment by ncw** [comment]
I'm sure I tried rclone with git-bash a while back and it worked properly.

I don't currently have git-bash though...

**Comment by zlorf** [comment]
I've tried old bundles of git-bash (e.g. 2.12) and the issue was still visible. I've tried also installed version (previously was using portable one) - no difference.

When you were using git-bash back then, was it based on mintty or Windows cmd console?

**Comment by ncw** [comment]
> When you were using git-bash back then, was it based on mintty or Windows cmd console?

I don't know, I'm afraid - I don't have the VM image any more.

I didn't install the whole of cygwin just the git-bash package - I remember that much.

It might be worth trying some other curses/VT100 programs and see whether they work.

From your screenshot I think rclone is doing some of the terminal commands it should be - do you agree?

These are the only escape sequences rclone uses

```
        EraseLine         = "\x1b[2K"
        MoveToStartOfLine = "\x1b[1G"
        MoveUp            = "\x1b[1A"
```

So maybe one of those isn't working properly.  You could try these which should look like this

```
$ echo -e "Hello \x1b[2K World"
       World
$ echo -e "Hello \x1b[1G World"
 World
$ echo -e "Hello\n\n \x1b[1A World"
Hello
  World
$ 
```

**Comment by zlorf** [comment]
I've tried other programs that should update output (e.g. curl with progress on downloading a big file) - they update output properly in place.

I've executed commands you've provided and it looks as expected:
![control-characters](https://user-images.githubusercontent.com/1769916/87096146-75e82680-c242-11ea-97e2-19d5ebd7d336.png)

While searching if someone has seen similar problem, I've stumbled upon https://github.com/rclone/rclone/issues/4322. It shows the same problem in - I guess - linux.
Just to confirm: the expected behavior is that `-P` shows progress that is constantly updating in place, right?

**Comment by ncw** [comment]
> I've executed commands you've provided and it looks as expected:

Great!



> Just to confirm: the expected behavior is that `-P` shows progress that is constantly updating in place, right?

That is correct.

The next possibility is that rclone isn't outputting the control codes at all. I wasn't sure from picture you posted.

There are 3 possibilities in the code, in order they are

1. Don't use escape sequences
2. Send escape sequences to the terminal
3. Use Windows CMD Windows calls to emulate the terminal movement on Windows or send escape sequences direct to the terminal

https://github.com/rclone/rclone/blob/0792f4722c2c9c62e45721b765f1a2a504af2158/lib/terminal/terminal.go#L75-L83

We've established that the escape sequences are working and you tell me `TERM` is set to `xterm` so I think this should be choosing option 2, unless by some chance the tty detection isn't working...

Can you put print statements in those branches and find out which one it is taking?

It would be worth trying [the latest beta](https://beta.rclone.org) also as I updated the terminal library that is depends on recently.

**Comment by ncw** [comment]
I made a beta with those print statements in for you - can you give it a go?

[v1.52.2-213-ge773b07d-fix-4415-term-beta](https://beta.rclone.org/branch/v1.52.2-213-ge773b07d-fix-4415-term-beta/) on branch [fix-4415-term](https://github.com/rclone/rclone/tree/fix-4415-term) (uploaded in 15-30 mins)

**Comment by zlorf** [comment]
Thank you so much for digging into the issue!
So I've downloaded and run the beta code and it prints `terminal: removing escape codes - no tty detected`.
That actually gave me a hint. Searching on internet I've found that prefixing rclone command with `winpty` solves the problem. Apparently, this is some limitation of mintty - e.g. https://github.com/git-for-windows/git/wiki/FAQ#some-native-console-programs-dont-work-when-run-from-git-bash-how-to-fix-it

So in the end: the solution to run rclone on git-bash with proper progress is to run:
```
winpty rclone copy [....]
```
